### PR TITLE
Added support for zend_oo data types

### PR DESCRIPTION
**Pull request by lander86**
Adding support for zend_oo data types on this branch.

### CODE REVIEW

File: lib/address.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ var errors = require('./errors');
 var Base58Check = require('./encoding/base58check');
 var Networks = require('./networks');
 var Hash = require('./crypto/hash');
-var JSUtil = require('./util/js');
```
Comment (otoumas) [inline lib/address.js:9]: Missing ;

**Review by paolocappelletti — DISMISSED** [review]

File: test/transaction/transaction-sidechain.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            assert.equal(sc.vft_ccout[i].address,txJson.vft_ccout[i].address);
+          }
+
+          //vmbtr_out
+          assert.equal(sc.vmbtr_out.length,txJson.vmbtr_out.length);
+          for (var i=0; i<txJson.vmbtr_out.length;i++) {
+              assert.equal(sc.vmbtr_out[i].scid,txJson.vmbtr_out[i].scid);
+              assert.equal(tx.sc_params.vmbtr_out[i].scFee / 1e8,txJson.vmbtr_out[i].scFee);
+              assert.equal(sc.vmbtr_out[i].mcDestinationAddress,txJson.vmbtr_out[i].mcDestinationAddress.pubkeyhash);
+              assert.equal(sc.vmbtr_out[i].vScRequestData.length,txJson.vmbtr_out[i].vScRequestData.length);
```
Comment (yurikoinaba) [inline test/transaction/transaction-sidechain.js:111]: Do you also want to check that the elements of the array are equal?

File: lib/transaction/transaction.js [hunk 3]
```diff
@@ -44,6 +44,12 @@ function Transaction(serialized) {
   this._inputAmount = undefined;
   this._outputAmount = undefined;
 
+  this.sc_params=undefined;
+  this.joinSplits = [];
```
Comment (yurikoinaba) [inline lib/transaction/transaction.js:48]: This is already defined 4 lines above.

File: lib/transaction/sidechain.js [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+    }
+}
+
+class Vft_ccout {
+    constructor (satoshis,satoshisBN, address, scid) {
+        this.satoshis = satoshis;
+        this.satoshisBN = satoshisBN;
+        this.address = address;
+        this.scid = scid;
+    }
```
Comment (MarcoOl94) [inline lib/transaction/sidechain.js:54]: add the new field **mcReturnAddress **

**Review by MarcoOl94 — CHANGES_REQUESTED** [review]

File: lib/transaction/transaction.js [hunk 5]
```diff
@@ -324,18 +411,33 @@ Transaction.prototype.fromBufferReader = function(reader) {
...
+  }else{
+    if(this.version == -4) {
+      this.sc_params=Sidechain.fromBufferReader(reader);
     }
-    if (sizeJSDescs > 0) {
-      this.joinSplitPubKey = reader.read(32);
-      this.joinSplitSig = reader.read(64);
+    if (this.version != -5) {
+      this.nLockTime = reader.readUInt32LE();
+    }
```
Comment (yurikoinaba) [inline lib/transaction/transaction.js:426]: `this.version != -5` will always be `true` because we are in the `else` block

File: lib/networks.js [hunk 6]
```diff
@@ -118,134 +118,96 @@ function addNetwork(data) {
...
+addNetwork({
+    name: 'regtest',
+    alias: 'dev',
+    pubkeyhash: 0x2098,
+    privatekey: 0xef,
+    scripthash: 0x2092,
+    bech32prefix: 'bcrt',
+    xpubkey: 0x043587cf,
+    xprivkey: 0x04358394,
+    networkMagic: 0x2f54cc9d,
```
Comment (zain-cheng) [inline lib/networks.js:189]: Just curious, what is networkMagic?

File: test/block/block.js [hunk 7]
```diff
@@ -47,25 +43,40 @@ describe('Block', function() {
...
+  it('should parse correctly transactions', function() {
+      var b = Block(blockbuf);
+      var txList = b.transactions;
+      txList.length.should.equal(2);
+      txList[0].id.should.equal(JSON.parse(dataJson).transactions[0].hash);
+      txList[1].id.should.equal(JSON.parse(dataJson).transactions[1].hash);
+  });
+
+
+/*
```
Comment (otoumas) [inline test/block/block.js:55]: There is a lot of commented code in this file, can we get rid of it?

File: test/block/blockheader.js [hunk 8]
```diff
@@ -4,32 +4,38 @@ var bitcore = require('../..');
...
+var data = require('../data/blk642579-testnet');
+var blockhex = data.blockhex;
+var blockbuf = new Buffer(blockhex, 'hex');
 
 describe('BlockHeader', function() {
 
   var version = data.version;
   var prevblockidbuf = new Buffer(data.prevblockidhex, 'hex');
   var merklerootbuf = new Buffer(data.merkleroothex, 'hex');
+  var hash = data.hash;
```
Comment (otoumas) [inline test/block/blockheader.js:24]: unused var

File: test/transaction/sidechain-certificate.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
+"use strict";
```
Comment (otoumas) [inline test/transaction/sidechain-certificate.js:1]: single quotes

File: test/transaction/sidechain-certificate.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+var assert = require('chai').assert;
+var testdataExpected =  require('../data_tx/tx-sidechain-certificate.json')
+var BufferUtil = require('../../lib/util/buffer');
+
+describe('#Sidechain certificate', function() {
+
+  describe('Certificate parsing', function() {
+    var index = 0;
+    testdataExpected.forEach(function (ele) {
+      it('vector #' + index+": "+ele.description, function() {
```
Comment (otoumas) [inline test/transaction/sidechain-certificate.js:13]: space after +

File: test/transaction/sidechain-certificate.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        assert.equal(cert.ftScFee / 1e8, certJson.cert.ftScFee);
+        //mbtrScFee
+        assert.equal(cert.mbtrScFee / 1e8, certJson.cert.mbtrScFee);
+
+        //vout
+        assert.equal(cert.outputs.length,certJson.vout.length);
+        for (var i=0; i<certJson.vout.length;i++) {
+          assert.equal(cert.outputs[i].toObject().satoshis,certJson.vout[i].valueZat);
+          assert.equal(cert.outputs[i].toObject().script,certJson.vout[i].scriptPubKey.hex);
+          if (cert.outputs[i].toObject().isFromBackwardTransfer && cert.outputs[i].toObject().isFromBackwardTransfer == true) {
```
Comment (otoumas) [inline test/transaction/sidechain-certificate.js:54]: use ===

File: test/transaction/sidechain-certificate.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+            assert.equal(cert.outputs[i].toObject().isFromBackwardTransfer,certJson.vout[i]["backward transfer"]);
```
Comment (otoumas) [inline test/transaction/sidechain-certificate.js:55]: single quote

File: test/transaction/transaction-sidechain.js [hunk 13]
```diff
@@ -0,0 +1,126 @@
+"use strict";
+
+var should = require("chai").should();
+var Transaction = require ("../../lib/transaction");
```
Comment (otoumas) [inline test/transaction/transaction-sidechain.js:4]: single quotes

File: test/transaction/transaction-sidechain.js [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+
+const revert = function(pubkey) {
+    var pubkeyhash = ""
+    var i = pubkey.length-1
+    while(i>0){
+        pubkeyhash+=pubkey[i-1]+pubkey[i]
+        i=i-2
+    }
+    return pubkeyhash
+}
```
Comment (otoumas) [inline test/transaction/transaction-sidechain.js:18]: ;

File: test/transaction/transaction-sidechain.js [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+
+describe('#Sidechain transactions', function() {
+
+  describe('SC_create/FT/VMBTR', function() {
+    var index = 0;
+    txes.forEach(function(ele) {
+
+      it('vector #' + index+": "+ele.description, function() {
```
Comment (otoumas) [inline test/transaction/transaction-sidechain.js:26]: space after +

**Review by otoumas — COMMENTED** [review]
A few comments 😅, most of them are just typos (missing ;, double/single quotes, === instead of ==)....
But in between all that noise there a re some open questions and posible changes to be made 😄

File: lib/block/blockheader.js [hunk 16]
```diff
@@ -27,7 +30,7 @@ var BlockHeader = function BlockHeader(arg) {
   this.version = info.version;
   this.prevHash = info.prevHash;
   this.merkleRoot = info.merkleRoot;
-  this.reserved = info.reserved;
+  this.scTxsCommitment = info.scTxsCommitment;
```
[the remark this answers is not in the document]
Comment (otoumas) [reply lib/block/blockheader.js:33]: nvm already solved this doubt with @MarcoOl94: 
version 4 keep reserved
version 3 change to scTxCommitment

File: lib/networks.js (hunk 6 above)
Comment (MarcoOl94) [reply lib/networks.js:189]: NetworkMagic are bits that are used as a way to identify the separate messages sent between nodes on the network.

File: lib/script/script.js [hunk 17]
```diff
@@ -420,39 +433,68 @@ Script.prototype.isScriptHashIn = function() {
...
+  return !!(this.chunks.length === 8 &&
+      this.chunks[0].opcodenum === Opcode.OP_DUP &&
+      this.chunks[1].opcodenum === Opcode.OP_HASH160 &&
+      this.chunks[2].buf &&
+      this.chunks[2].buf.length === 20 &&
+      this.chunks[3].opcodenum === Opcode.OP_EQUALVERIFY &&
+      this.chunks[4].opcodenum === Opcode.OP_CHECKSIG  &&
+      this.chunks[5].buf &&
+      this.chunks[5].buf.length === 32 &&
+      //this.chunks[6].opcodenum === Opcode.OP_SMALLDATA &&
```
[the remark this answers is not in the document]
Comment (otoumas) [reply lib/script/script.js:470]: chunk[6] should be an integer, we can either find an OP_code representing this or delete the line.

File: lib/script/script.js [hunk 18]
```diff
@@ -697,17 +732,18 @@ Script.prototype.removeCodeseparators = function() {
...
   publicKeys = _.map(publicKeys, PublicKey);
   var sorted = publicKeys;
-  if (!opts.noSorting) {
+  //Zend don't sort the pubkey
+  /*if (!opts.noSorting) {
     sorted = _.sortBy(publicKeys, function(publicKey) {
       return publicKey.toString('hex');
     });
-  }
+  }*/
```
[the remark this answers is not in the document]
Comment (otoumas) [reply lib/script/script.js:746]: Add to the datadoc of the method this difference with bitcoin: In zend the public keys are not sorted in multisig scripts

File: lib/script/script.js [hunk 19]
```diff
@@ -930,6 +966,65 @@ Script.prototype.getAddressInfo = function(opts) {
...
+  return info;
+};
+
+
+/**
+ * @returns {Script} a new OP_RETURN script with data
+ * @param {(string|Buffer)} data - the data to embed in the output
+ * @param {(string)} encoding - the type of encoding of the string
+ */
+Script.buildDataOut = function(data, encoding) {
```
[the remark this answers is not in the document]
Comment (otoumas) [reply lib/script/script.js:997]: Investigate it, the function is being called by transaction.js line 987 T`ransaction.prototype.addData = function(value) {`

File: lib/script/script.js [hunk 20]
```diff
@@ -930,6 +966,65 @@ Script.prototype.getAddressInfo = function(opts) {
...
+
+/**
+ * Will return the associated address information object
+ * @return {Address|boolean}
+ */
+Script.prototype.getAddressInfo = function(opts) {
+  if (this._isInput) {
+    return this._getInputAddressInfo();
+  } else if (this._isOutput) {
+    return this._getOutputAddressInfo();
```
[the remark this answers is not in the document]
Comment (otoumas) [reply lib/script/script.js:1018]: This attributes are set in output.js and in input.js

File: lib/transaction/input/multisig.js [hunk 21]
```diff
@@ -23,7 +23,9 @@ function MultiSigInput(input, pubkeys, threshold, signatures) {
   pubkeys = pubkeys || input.publicKeys;
   threshold = threshold || input.threshold;
   signatures = signatures || input.signatures;
-  this.publicKeys = _.sortBy(pubkeys, function(publicKey) { return publicKey.toString('hex'); });
+  this.publicKeys = pubkeys.map((pubkey) => pubkey.toString('hex'));
+  //Zend don't sort the pubkey
+  //this.publicKeys = _.sortBy(pubkeys, function(publicKey) { return publicKey.toString('hex'); });
```
[the remark this answers is not in the document]
Comment (otoumas) [reply lib/transaction/input/multisig.js:28]: Add comment to the method doc as this is a difference with Bitcoin

File: lib/transaction/output.js [hunk 22]
```diff
@@ -29,11 +31,34 @@ function Output(args) {
...
+* The transaction that will create this kind of output will be responsible to set this flag.
+*/
+Object.defineProperty(Output.prototype, 'isFromBackwardTransfer', {
+  configurable: false,
+  enumerable: true,
+  get: function() {
+     return this._isFromBackwardTransfer ? this._isFromBackwardTransfer : false;
+  },
+  set: function(val) {
+    return this._isFromBackwardTransfer = val;
```
[the remark this answers is not in the document]
Comment (otoumas) [reply lib/transaction/output.js:58]: No, this is correct as it is a set function. We need to remove the return keyword

File: lib/transaction/sidechain.js [hunk 23]
```diff
@@ -0,0 +1,469 @@
...
+  var vcsw=[];
+  _.each(params.vcsw_ccin, function(vcsw_ccin){
+    var values = new Sidechain._checkAmount(vcsw_ccin.value);
+    vcsw.push(new Vcsw_ccin(
+        values[0],
+        values[1],
+        BufferUtil.reverse(new buffer.Buffer(vcsw_ccin.scId, 'hex')),
+        new Buffer(vcsw_ccin.nullifier, 'hex'),
+        BufferUtil.reverse(new Buffer(vcsw_ccin.pubKeyHash, 'hex')),
+        new buffer.Buffer(vcsw_ccin.scProof, 'hex'),
```
[the remark this answers is not in the document]
Comment (otoumas) [reply lib/transaction/sidechain.js:97]: unify to new buffer.Buffer

File: lib/transaction/sidechain.js [hunk 24]
```diff
@@ -0,0 +1,469 @@
...
+    certVk = br.read(certVk_len);
+    wCeasedVk_option = br.readVarintNum();
+    if (wCeasedVk_option == 1) {
+      wCeasedVk_len = br.readVarintNum();
+      wCeasedVk = br.read(wCeasedVk_len);
+    }
+    vFieldElementCertificateFieldConfig_len = br.readVarintNum();
+    for (j = 0; j < vFieldElementCertificateFieldConfig_len; j++) {
+      vFieldElementCertificateFieldConfig.push(br.readUInt8());
+    }
```
[the remark this answers is not in the document]
Comment (otoumas) [reply lib/transaction/sidechain.js:254]: Nope

File: lib/transaction/transaction.js [hunk 25]
```diff
@@ -111,6 +117,20 @@ var hashProperty = {
 Object.defineProperty(Transaction.prototype, 'hash', hashProperty);
 Object.defineProperty(Transaction.prototype, 'id', hashProperty);
 
+
+Transaction.prototype.getValueOfBackwardTransfers = function() {
```
[the remark this answers is not in the document]
Comment (otoumas) [reply lib/transaction/transaction.js:121]: Used in block explorer

File: lib/transaction/transaction.js [hunk 26]
```diff
@@ -375,6 +502,8 @@ Transaction.prototype.toObject = Transaction.prototype.toJSON = function toObjec
   if (!_.isUndefined(this._fee)) {
     obj.fee = this._fee;
   }
+  obj.hex=this.toBuffer().toString('hex');
+
   return obj;
 };
```
[the remark this answers is not in the document]
Comment (otoumas) [reply lib/transaction/transaction.js:509]: Add a test and do the change

**Comment by zain-cheng** [comment]
From a high level, this looks fine to me.  I haven't gotten a chance to dig deeper.  @paolocappelletti Are you ready to approve the PR? Let me know if we need two approvers.

**Review by paolocappelletti — APPROVED** [review]

**Review by zain-cheng — APPROVED** [review]
@otoumas @MarcoOl94 I'm okay with merging these changes and then iterating on it.

**Pull request merged**